roachtest: sqlsmith failure could not find overload for st_dwithinexclusive

[(roachtest).sqlsmith/setup=seed/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2374262&tab=buildLog) on [master@ab503e2fd708541e5e9ebb9a6f2651eda506f2ef](https://github.com/cockroachdb/cockroach/commits/ab503e2fd708541e5e9ebb9a6f2651eda506f2ef):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=seed/setting=no-mutations/run_1
	sqlsmith.go:181,sqlsmith.go:211,test_runner.go:755: error: pq: internal error: could not find overload for st_dwithinexclusive
		stmt:
		SELECT
			'\x27':::BYTES AS col_36252,
			tab_14679._int2 AS col_36253,
			tab_14678.tableoid AS col_36254,
			tab_14678._inet AS col_36255,
			tab_14678._timestamp AS col_36256,
			tab_14679._string AS col_36257,
			tab_14679._bool AS col_36258,
			tab_14679._bool AS col_36259,
			7232820974290837299:::INT8 AS col_36260,
			tab_14678._enum AS col_36261,
			tab_14679.tableoid AS col_36262,
			0.8972772296191472:::FLOAT8 AS col_36263,
			tab_14678._interval AS col_36264,
			(-32768):::INT8 AS col_36265,
			tab_14679._jsonb AS col_36266
		FROM
			defaultdb.public.seed@[0] AS tab_14678, defaultdb.public.seed@[0] AS tab_14679
		WHERE
			((-0.7465414041412649):::FLOAT8 <= st_distance(st_geogfromwkb(tab_14678._bytes::BYTES, tab_14678._int2::INT8)::GEOGRAPHY::GEOGRAPHY, '0105000020E61000000100000001020000000600000034C5D88D6BE861404C4234EC259A55C0ACE2F00BFF78624044FFEB6C8C3030C038E263FE39934A40C43496FF73643340F0DABC4503CD3C40643A74D2576E48409421A4C9142842C0C04A7A832A340F40DD59E540C7AB63C0ACEA15A4216734C0':::GEOGRAPHY::GEOGRAPHY, tab_14678._bool::BOOL)::FLOAT8)
		LIMIT
			89:::INT8;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=seed/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=2374262&tab=artifacts#/sqlsmith/setup=seed/setting=no-mutations)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Dseed%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/opt/norm/comp_funcs.go
+++ b/pkg/sql/opt/norm/comp_funcs.go
@@ -237,8 +237,18 @@ func (c *CustomFuncs) makeSTDWithin(
 		}
 	}
 
+	newArgs := make(memo.ScalarListExpr, len(args)+1)
+	const distanceIdx, useSpheroidIdx = 2, 3
+	copy(newArgs, args[:distanceIdx])
+
 	// The distance parameter must be type float.
-	newArgs := append(args, c.f.ConstructCast(bound, types.Float))
+	newArgs[distanceIdx] = c.f.ConstructCast(bound, types.Float)
+
+	// Add the use_spheroid parameter if it exists.
+	if len(newArgs) > useSpheroidIdx {
+		newArgs[useSpheroidIdx] = args[useSpheroidIdx-1]
+	}
+
 	props, overload, ok := memo.FindFunction(&newArgs, name)
 	if !ok {
 		panic(errors.AssertionFailedf("could not find overload for %s", name))

--- a/pkg/sql/opt/norm/testdata/rules/comp
+++ b/pkg/sql/opt/norm/testdata/rules/comp
@@ -875,6 +875,18 @@ select
                 └── cast: INT8 [type=int]
                      └── variable: val:3 [type=float]
 
+# Regression test for #55675. Handle use_spheroid param.
+norm expect=FoldCmpSTDistanceLeft
+SELECT * FROM geom_geog WHERE st_distance(geog, 'point(0.0 0.0)', true) <= 5
+----
+select
+ ├── columns: geom:1 geog:2!null val:3
+ ├── immutable
+ ├── scan geom_geog
+ │    └── columns: geom:1 geog:2 val:3
+ └── filters
+      └── st_dwithin(geog:2, '0101000020E610000000000000000000000000000000000000', 5.0, true) [outer=(2), immutable, constraints=(/2: (/NULL - ])]
+
 # --------------------------------------------------
 # FoldCmpSTDistanceRight
 # --------------------------------------------------
@@ -926,6 +938,18 @@ select
  │    └── columns: geom:1 geog:2 val:3
  └── filters
       └── st_dwithinexclusive(geom:1, '010100000000000000000000000000000000000000', val:3) [outer=(1,3), immutable, constraints=(/1: (/NULL - ]; /3: (/NULL - ])]
+
+# Regression test for #55675. Handle use_spheroid param.
+norm expect=FoldCmpSTDistanceRight
+SELECT * FROM geom_geog WHERE val > st_distance(geog, 'point(0.0 0.0)', false)
+----
+select
+ ├── columns: geom:1 geog:2!null val:3!null
+ ├── immutable
+ ├── scan geom_geog
+ │    └── columns: geom:1 geog:2 val:3
+ └── filters
+      └── st_dwithinexclusive(geog:2, '0101000020E610000000000000000000000000000000000000', val:3, false) [outer=(2,3), immutable, constraints=(/2: (/NULL - ]; /3: (/NULL - ])]
 
 # --------------------------------------------------
 # FoldCmpSTMaxDistanceLeft

